src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -119,26 +118,17 @@ void AdaptDispatch::_WriteToBuffer(const std::wstring_view string)
             }
         }
 
-        if (_modes.test(Mode::InsertReplace))
-        {
-            // If insert-replace mode is enabled, we first measure how many cells
-            // the string will occupy, and scroll the target area right by that
-            // amount to make space for the incoming text.
-            const OutputCellIterator it(state.text, attributes);
-            auto measureIt = it;
-            while (measureIt && measureIt.GetCellDistance(it) < state.columnLimit)
-            {
-                ++measureIt;
-            }
-            const auto row = cursorPosition.y;
-            const auto cellCount = measureIt.GetCellDistance(it);
-            _ScrollRectHorizontally(textBuffer, { cursorPosition.x, row, state.columnLimit, row + 1 }, cellCount);
-        }
-
         state.columnBegin = cursorPosition.x;
 
         const auto textPositionBefore = state.text.data();
-        textBuffer.Write(cursorPosition.y, attributes, state);
+        if (_modes.test(Mode::InsertReplace))
+        {
+            textBuffer.Insert(cursorPosition.y, attributes, state);
+        }
```

Comment:
This isn't working correctly when you have horizontal margins set. If you're inserting text while inside the margins, the content that is pushed to the right should only move as far as the right margin. I didn't look at the implementation that closely, but I'm guessing the `Insert` method is not taking the `state.columnLimit` into account.

Reply:
Test case in pwsh:
```
"`e7`e[?69h`e[1;20s`e8*`e[19b`r`e[4h `e[19b`e[4l`e[s`e8`n"
```
That should produce what is effectively a blank line, but instead we're seeing 20 `*` characters offset by 20 columns.

Reply:
I believe I fixed this issue. Thanks!

I noticed that after running your test string, `TextBuffer::SetCurrentLineRendition` won't get called anymore. Is that to be expected? (It does work again after emitting a RIS.)

Reply:
I didn't think about that, but that is actually correct. When the horizontal margin mode is enabled, line renditions are disabled. I should have reset mode ?69 at the end of that test case.

Reply:
@lhecker